### PR TITLE
fix when reclustering jets and MET on the fly with postfix option

### PR DESCRIPTION
**Pull request by gouskos**
This PR fixes an issue with the MET tool when trying to recluster jets/MET on the fly. 
It does not change the data content

Package modified: PhysicsTools/PatUtils/

### CODE REVIEW

**Comment by perrotta** [comment]
please test

File: PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py [hunk 1]
```diff
@@ -1443,8 +1448,10 @@ def ak4JetReclustering(self,process, pfCandCollection, patMetModuleSequence, pos
             pfCHS=None
             if self._parameters["onMiniAOD"].value: 
                 pfCHS = cms.EDFilter("CandPtrSelector", src = pfCandCollection, cut = cms.string("fromPV"))
-                setattr(process,"pfNoPileUpJME"+postfix,pfCHS)
+                #setattr(process,"pfNoPileUpJME"+postfix,pfCHS)
```
Comment (slava77) [inline PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py]: is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant

**Comment by slava77** [comment]
> This PR fixes an issue with the MET tool when trying to recluster jets/MET on the fly.

@gouskos 
IIUC, this is a fix to some setup that is not used in production miniAOD or CMSSW pat unit tests.
please point to an example of what was broken that is now fixed with this PR.
Thank you.

File: PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py (hunk 1 above)
Comment (gouskos) [reply PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py]: @slava77  No is not - sorry. We fixed it with a new commit.

**Comment by gouskos** [comment]
> IIUC, this is a fix to some setup that is not used in production miniAOD or CMSSW pat unit tests.
please point to an example of what was broken that is now fixed with this PR.
Thank you.

The fix in this PR is more relevant to users that would like to do more involved studies with MET. One example is to remove candidates in HF from the MET computation. Jets are rebuilt from those candidates by the tools, no need to do anything else.

Please find in this link [test_metrecipes_cfg.py](https://gist.github.com/gouskos/a5252e91f82ba349d46eb733f85c00d6) a working example. Running this code before the fix would give the following error:

> An exception of category 'Invalid Jet Inputs' occurred while
   [0] Processing  Event run: 1 lumi: 162 event: 16130 stream: 0
   [1] Running path 'p'
   [2] Calling method for module FastjetJetProducer/'ak4PFJetsCHSTest'

i.e. It does not recognize the new jet collection if a "PostFix" label is used.

**Comment by slava77** [comment]
On 2/22/18 10:24 AM, gouskos wrote:
>     IIUC, this is a fix to some setup that is not used in production
>     miniAOD or CMSSW pat unit tests.
>     please point to an example of what was broken that is now fixed with
>     this PR.
>     Thank you.
> 
> The fix in this PR is more relevant to users that would like to do more
> involved studies with MET. One example is to remove candidates in HF
> from the MET computation. Jets are rebuilt from those candidates by the
> tools, no need to do anything else.
> 
> Please find in this link test_metrecipes_cfg.py
> <https://gist.github.com/gouskos/a5252e91f82ba349d46eb733f85c00d6> a
> working example. Running this code before the fix would give the
> following error:
> 
>     An exception of category 'Invalid Jet Inputs' occurred while
>     [0] Processing Event run: 1 lumi: 162 event: 16130 stream: 0
>     [1] Running path 'p'
>     [2] Calling method for module FastjetJetProducer/'ak4PFJetsCHSTest'
> 
> i.e. It does not recognize the new jet collection if a "PostFix" label
> is used.
> 

Thanks.

Considering that the issue related to HF candidates is supposed to enter
the production code for 80X legacy re-miniAOD, it's not just a fix for
users for some corner cases.

**Comment by gouskos** [comment]
Hi Slava. The noHFMET was just an example. We would like to fix this functionality of tool in case is needed.

**Comment by perrotta** [comment]
please test

File: PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py [hunk 2]
```diff
@@ -452,6 +453,10 @@ def produceMET(self, process,  metType, metModuleSequence, postfix):
         if postfix != "" and metType == "PF" and not hasattr(process, 'pat'+metType+'Met'+postfix):
             noClonesTmp = [ "particleFlowDisplacedVertex", "pfCandidateToVertexAssociation" ]
             configtools.cloneProcessingSnippet(process, getattr(process,"producePatPFMETCorrections"), postfix, noClones = noClonesTmp, addToTask = True)
+            #MM FIXME, this could be done in a mch better way, this is not
+            #handled by the above cloning sequence
+            getattr(process,"selectedPatJetsForMetT1T2Corr").src=jetCollectionUnskimmed #cms.InputTag("patJets"+postfix)
```
Comment (slava77) [inline PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py]: the result of this call does not quite make sense to me. In the standard config I see the following:
- selectedPatJetsForMetT1T2Corr has src = slimmedJetsPuppi
- selectedPatJetsForMetT1T2CorrPuppi has patJetsNoHF
- selectedPatJetsForMetT1T2CorrNoHF has (unchanged) patJets

All of the above had patJets in the baseline setup.

Comment (slava77) [reply PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py]: these modules are not running in the default (mini)AOD configuration.
Is there something else missing in the configuration that would enable them or are they instantiated for no particular reason?

**Comment by slava77** [comment]
@gouskos 
I used your file https://gist.github.com/gouskos/a5252e91f82ba349d46eb733f85c00d6/raw/269391c78896266b6cba12ac442eb115f3ddffa4/test_metrecipes_cfg.py

I confirm that it fails in the baseline and runs OK with this PR.
If I understand correctly, the error ``` Did not specify appropriate inputs for VirtualJetProducer, Abort!``` from ak4PFJetsCHSTest is fixed appropriately with the sequence/task manipulations in ak4JetReclustering.

However, in the expanded config I see a somewhat similar oddity as in the standard config that I have described in https://github.com/cms-sw/cmssw/pull/22273#discussion_r171077669
- selectedPatJetsForMetT1T2Corr has its input changed to patJetsTest. I think that it should stay unchanged
- selectedPatJetsForMetT1T2CorrTest input changed from patJets to patJetsTest. OK
- selectedPatJetsForMetT2Corr input changed from patJets to patJetsTest. I think that it should stay unchanged
- selectedPatJetsForMetT2CorrTest input did not change and remained patJets. I think that this should actually change.

It seems to me that the change that starts with ```#MM FIXME``` should not be done for now because it looks similarly "broken" as the original.

**Comment by gouskos** [comment]
Thank you @slava77 . We will test it by undoing the change with #MM fixme and keep you posted.

**Comment by slava77** [comment]
@gouskos 
please clarify on the status of updates to this PR
Thank you.

**Comment by gouskos** [comment]
Hi @slava77 . A small update from our side: 

While checking the details we found a bug in  [https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/tools/helpers.py#L121-L134](url)

The code after the fix should read:
```
def removeIfInSequence(process, target,  sequenceLabel, postfix=""):
    labels = __labelsInSequence(process, sequenceLabel, postfix, True)                                                                                                                        
    if target+postfix in labels:
        getattr(process, sequenceLabel+postfix).remove(
            getattr(process, target+postfix)
            )

def __labelsInSequence(process, sequenceLabel, postfix="", keepPostFix=False):                                                                                                                     
    position = -len(postfix)
    if keepPostFix:
        position = None

    result = [ m.label()[:position] for m in listModules( getattr(process,sequenceLabel+postfix))]
    result.extend([ m.label()[:position] for m in listSequences( getattr(process,sequenceLabel+postfix))] ) getattr(process,sequenceLabel+postfix))] )                                                                             
    if postfix == "":
        result = [ m.label() for m in listModules( getattr(process,sequenceLabel+postfix))]
        result.extend([ m.label() for m in listSequences( getattr(process,sequenceLabel+postfix))]  )
    return result
```

After this fix, some modifications introduced in runMETCorrectionsAndUncertainties.py [which you also comment #22273 are not necessary [i.e. we can remove these lines]]. We are doing some final tests, but if this seems reasonable to you we will update the PR.

**Comment by slava77** [comment]
> We are doing some final tests, but if this seems reasonable to you we will update the PR.

visually, this appears reasonable.
Please go ahead with the update.

**Comment by gouskos** [comment]
Hi @slava77  the IB we used for this PR is not available anymore and I cannot push the changes using the latest IB [complains that the fixmettool_1010 branch is behind the current branch]. Are there any instructions on how to do it? In addition I would like to checkout a new package [PhysicsTools/PatAlgos] to fix the bug discussed earlier. Thank you.

**Comment by gouskos** [comment]
Sorry for the spam @slava77 . I think is described here [http://cms-sw.github.io/tutorial-resolve-conflicts.html#the-above-is-all-great-stuff-but-i-need-a-quick-recipe](url)

**Comment by slava77** [comment]
right, http://cms-sw.github.io/tutorial-resolve-conflicts.html
should help

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@gouskos 

I can confirm that now in the test_metrecipes_cfg.py the Sequence patPFMetT1T2CorrSequenceTest no longer holds selectedPatJetsForMetT1T2CorrTest, which is intended to be removed by  https://github.com/cms-met/cmssw/blob/b9c397c040c57018bef51d43c604cd5d570de821/PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py#L431
.. and this apparently worked already even in the baseline for the standard setup of miniAOD (e.g. wf 136.7611)

going back to the previously proposed diff from  PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py 
```diff
+            getattr(process,"selectedPatJetsForMetT1T2Corr").src=jetCollectionUnskimmed #cms.InputTag("patJets"+postfix)
+            getattr(process,"selectedPatJetsForMetT2Corr").src=jetCollectionUnskimmed #cms.InputTag("patJets"+postfix)
```

Now we have selectedPatJetsForMetT1T2Corr, selectedPatJetsForMetT1T2CorrNoHF, and selectedPatJetsForMetT1T2CorrPuppi all point to the same patJets as a source; the same as in the baseline.
Is it still intentional?
Wasn't the goal to set patJets, patJetsNoHF and patJetsPuppi for these modules, respectively?
Should the change in runMETCorrectionsAndUncertainties be actually reinstated with 
```diff
+            getattr(process,"selectedPatJetsForMetT1T2Corr").src=cms.InputTag("patJets"+postfix)
+            getattr(process,"selectedPatJetsForMetT2Corr").src=cms.InputTag("patJets"+postfix)
```
?
I guess we can sort that out in a separate PR.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/22273	b9c397c
- insert tmpPFCandCollPtrNoHF and pfNoPileUpJMENoHF in the sequence definition for appropriate top projection of pf candidates; appropriately actually drop selectedPatJetsForMetT1T2Corr+postfix for all (more special) cases.
- jenkins tests pass and comparisons show no differences
- local checks of expanded configs show that there is no effect on the standard miniAOD workflows; the previously crashing test_metrecipes_cfg is now completing OK

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
merge
